### PR TITLE
IR: turn some IR nodes into data classes

### PR DESCRIPTION
**Pull request by ekiwi**
As discussed with @jackkoenig , I am changing Stop, Print and Verification to be a data class.
This is done in preparation for a new `name` field in order to preserve backwards compatibility.

### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- code refactoring

#### API Impact

- it should maintain the current API

#### Backend Code Generation Impact
- none

#### Desired Merge Strategy

- squash

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (1.2.x, 1.3.0, 1.4.0) ?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you mark as `Please Merge`?

### CODE REVIEW

**Review by chick — APPROVED** [review]
This looks pretty straightforward to me. 
The @data provided constructor seems to confuse IntelliJ a bit but things compile and run.
You should probably wait for second review from @jackkoenig before merging

**Comment by ekiwi** [comment]
>  The @DaTa provided constructor seems to confuse IntelliJ a bit but things compile and run.

Yes. That is somewhat annoying.



>  You should probably wait for second review from @jackkoenig before merging

I especially would like @jackkoenig to check if there is a way to simplify the additions to the `build.sbt`.

**Comment by jackkoenig** [comment]
As an experiment, I went ahead and backported this and checked binary compatibility against 1.4.2, here were the errors:
```
firrtl: Failed binary compatibility check against edu.berkeley.cs:firrtl_2.12:1.4.2! Found 12 potential problems
 * class firrtl.ir.Print is declared final in current version
   filter with: ProblemFilters.exclude[FinalClassProblem]("firrtl.ir.Print")
 * static method tupled()scala.Function1 in class firrtl.ir.Print does not have a correspondent in current version
   filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("firrtl.ir.Print.tupled")
 * static method curried()scala.Function1 in class firrtl.ir.Print does not have a correspondent in current version
   filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("firrtl.ir.Print.curried")
 * the type hierarchy of object firrtl.ir.Print is different in current version. Missing types {scala.runtime.AbstractFunction5}
   filter with: ProblemFilters.exclude[MissingTypesProblem]("firrtl.ir.Print$")
< and the same for Verification and Stop >
```
These are all good overall, and it should be binary compatible for adding new fields going forward, but it's a bit of a bummer that we can't easily switch to `@data` on a minor version without breaking binary compatibility.

File: build.sbt
Comment (jackkoenig) [inline build.sbt:51]:
```suggestion
    "io.github.alexarchambault" %% "data-class" % "0.2.5",
```

File: src/main/scala/firrtl/ir/IR.scala
Comment (jackkoenig) [inline src/main/scala/firrtl/ir/IR.scala:616]:
```suggestion
  def unapply(s: Stop): Some[(Info, Int, Expression, Expression)] = {
```

This is not something I knew until just today I happened to see a conversation about it on Gitter. In Scala 2.13.4 you'll get an exhaustivity warning unless you return the more specific type: See https://github.com/scodec/scodec/pull/296 and the linked Gitter conversation.

File: src/main/scala/firrtl/ir/IR.scala
Comment (jackkoenig) [inline src/main/scala/firrtl/ir/IR.scala:650]:
```suggestion
  def unapply(s: Print): Some[(Info, StringLit, Seq[Expression], Expression, Expression)] = {
```

File: src/main/scala/firrtl/ir/IR.scala
Comment (jackkoenig) [inline src/main/scala/firrtl/ir/IR.scala:695]:
```suggestion
  def unapply(s: Verification): Some[(Formal.Value, Info, Expression, Expression, Expression, StringLit)] = {
```

**Review by jackkoenig — CHANGES_REQUESTED** [review]
build.sbt changes look good; it's the best way to express macros in 2.13 and 2.12 that I've come across.

I have a few requested changes though.

**Comment by davidbiancolin** [comment]
Is there an issue that provides more context for this? Could it be linked in the PR?

**Comment by ekiwi** [comment]
> Is there an issue that provides more context for this? Could it be linked in the PR?

This PR followed from a discussion we had on the UCB-BAR slack in the `chisel-dev` channel.
The most important advantage of `data` classes vs. `case` classes is that `data` classes do not define `unapply` and `copy` methods which are hard/impossible to keep backwards compatible when adding a new field.

**Comment by jackkoenig** [comment]
> Is there an issue that provides more context for this? Could it be linked in the PR?

I'm not sure that there is so we might as well make it here.

Basically, case classes are terrible for binary compatibility if you need to add another field. Adding a field breaks binary compatibility for the `copy` and `unapply` methods. I believe if you implement the `copy` method yourself, the Scala compiler will *not* generate one for you. I think the same is supposed to occur for `unapply`, but I think that is broken in Scala 2.12 despite release notes for [Scala 2.12.12 saying it should work](https://github.com/scala/scala/releases/tag/v2.12.2), here's proof showing it doesn't work in 2.12.12 https://scastie.scala-lang.org/MlNqn8ClQWalnUuZGuZocw (h/t @ekiwi).

**Comment by davidbiancolin** [comment]
Great, thanks @ekiwi & @jackkoenig. And this is in preparation for #2057, which presumably would make it possible to annotate those statements? (Sorry for the tangent.)

**Comment by jackkoenig** [comment]
> Great, thanks @ekiwi & @jackkoenig. And this is in preparation for #2057, which presumably would make it possible to annotate those statements? (Sorry for the tangent.)

Exactly. I think there's a debate to be had if we should use `@data` for everything, or only opportunistically whenever we want to add fields. This is at least letting Kevin do what he wants to accomplish while we mull over the larger policy decision.

One very unfortunate thing is that Scala 3 does not support macro annotations so we will have to use a different approach once we get there, but since the whole "unapply" generation thing is fixed in Scala 2.13, it's probably fixed in Scala 3 as well so presumably we can just use case classes with careful migration.

**Review by jackkoenig — APPROVED** [review]
LGTM

**Pull request merged**
